R/translator.R

```diff
@@ -0,0 +1,143 @@
+#'
+#' @return
+#' @export
+#'
+#' @import jsonlite
+#'
+#' @examples
+#' @rdname translator
+Translator <- setRefClass(
+  "Translator",
+  fields = list(
+    languages = "character",
+    key_translation = "character",
+    translation_language = "character",
+    translation_file = "character",
+    options = "list",
+    translations = "data.frame",
+    mode = "character"
+  )
+)
+
+#' @rdname translator
+Translator$methods(
+  initialize = function(translation_file, key_translation) {
+    translation_file <<- translation_file
+    key_translation <<- key_translation
+    options <<- .translator_options
+    # determine the translations format
+    translation_split <- unlist(strsplit(translation_file, "[.]"))
+    translation_format <- translation_split[length(translation_split)]
+    if (length(translation_split) == 1)
+      .read_csv(translation_file, key_translation)
+    else if (translation_format == "json")
+      .read_json(translation_file, key_translation)
+    else
+      stop("Unrecognized format of the file.")
+    # Set default translation language
+    translation_language <<- character(0)
+  },
+  .read_json = function(translation_file, key_translation) {
+    mode <<- "json"
+    # TODO validate format of json translation_file
+    # Update the list of options, or take a default from config.
+    json_data <- jsonlite::fromJSON(translation_file)
+    common_fields <- intersect(names(json_data), names(options))
+    options <<- modifyList(options, json_data[common_fields])
+
+    # If key_translation is not found translations, pick the first one from the list.
+    languages <<- as.vector(json_data$languages)
+    key_translation <<- check_value_presence(
+      key_translation,
+      languages,
+      sprintf("Translation '%s' not found, '%s' set as default",
+              key_translation, languages[1])
+      )
+
+    # read translations
+    translations <<- column_to_row(json_data$translation, key_translation)
+  },
+  .read_csv = function(translation_file, key_translation) {
+    mode <<- "csv"
+    # Config setting loading
+    translation_csv_config <- paste0(translation_file, ".yaml")
+    if (file.exists(translation_csv_config)){
+      local_config <- yaml.load_file(translation_csv_config)
+      options <<- modifyList(options, local_config)
+    }
+    else
+      warning(paste0("You didn't specify config translation yaml file. ",
+                     "Default settings are used."))
+
+    # read all csv files in format: 'translation_file_<language_code>'
+    translation_files <- list.files(dirname(translation_file))
+    translation_files <- translation_files[grep(paste0(translation_file,
+                                                       "_[[:lower:]]{2}[.]csv"),
+                                                       translation_files)]
+    tmp_translation <- multmerge(translation_files)
+    languages <<- as.vector(colnames(tmp_translation))
+    key_translation <<- check_value_presence(
```

Comment:
1. do we need the "key_translation" at all in the CSV case? It's hard for me to tell before I see CSV example of translations
2. it seems that key_translation is only needed in the json case. If that's right, then having the "key_translation" as an argument to initialize and used everywhere is unnecessary and complicates the code for little gain. Maybe this is an artifact of the original JSON-based approach? If key_translation is JSON-specific, maybe it should be declared in the JSON translation file, and not as an argument to initialize?